dowhy/gcm/anomaly.py

```diff
@@ -0,0 +1,186 @@
+    noise_dependent_function, nodes_order = get_noise_dependent_function(causal_model, target_node)
+    anomaly_scorer.fit(node_samples[target_node].to_numpy())
+
+    attributions = estimate_anomaly_score_attributions(noise_of_anomaly_samples[nodes_order].to_numpy(),
+                                                       noise_samples[nodes_order].to_numpy(),
+                                                       lambda x: anomaly_scorer.score(noise_dependent_function(x)),
+                                                       attribute_mean_deviation,
+                                                       shapley_config)
+
+    return {node: attributions[:, i] for i, node in enumerate(nodes_order)}
+
+
+def estimate_anomaly_score_attributions(samples_of_interest: np.ndarray,
+                                        distribution_samples: np.ndarray,
+                                        anomaly_scoring_func: Callable[[np.ndarray], np.ndarray],
+                                        attribute_mean_deviation: bool,
+                                        shapley_config: Optional[ShapleyConfig] = None) -> np.ndarray:
+    """
+    Estimates the contributions of the features for each sample in samples_of_interest to the anomaly score obtained by
+    the anomaly_scoring_func. If attribute_mean_deviation is set to False, the anomaly score is based on the information
+    theoretic  (IT) score -log(P(g(X) >= g(x))), where g is the anomaly_scoring_func, X samples from the marginal
+    distribution of the target_node and x an observation of the target_node in anomaly_samples. If
+    attribute_mean_deviation is set to True, the contribution to g(x) - E[g(X)] is estimated instead, i.e. the
+    feature relevance for the given scorer.
+
+    Note that the anomaly scoring function needs to handle the dimension and modality of the data. An example for a
+    function for multidimensional continues data would be:
+        density_estimator = GaussianMixtureDensityEstimator()
+        density_estimator.fit(original_observations)
+        anomaly_scoring_func = lambda x, y: estimate_inverse_density_score(x, y, density_estimator)
+
+    Related paper:
+    Janzing, D., Budhathoki, K., Minorics, L., & Bloebaum, P. (2019).
+    Causal structure based root cause analysis of outliers
+    https://arxiv.org/abs/1912.02724
+
+    :param samples_of_interest: Samples x for which the contributions are estimated.
+    :param distribution_samples: Samples from the (non-anomalous) distribution X.
+    :param anomaly_scoring_func: A function g that takes a sample from X as input and returns an anomaly score.
+    :param attribute_mean_deviation: If set to False, the contribution is estimated based on the IT score and if it is
+    set to True, the contribution is based on the feature relevance with respect to the given scoring function.
+    :param shapley_config: Config for the Shapley estimator.
```

Comment:
Will be good to point to the Shapley file/class to know how to create this config

Reply:
Added a doc reference.